modules/table/src/test/java/org/apache/ignite/internal/table/type/NumericTypesSerializerTest.java

```diff
@@ -0,0 +1,203 @@
+            }
+        });
+
+    }
+
+    /**
+     *
+     */
+    @ParameterizedTest
+    @MethodSource("numbers")
+    public void testNumber(Pair<BigInteger, BigInteger> pair) {
+        schema = new SchemaDescriptor(
+            UUID.randomUUID(),
+            42,
+            new Column[] {new Column("key", NativeTypes.INT64, false)},
+            new Column[] {
+                new Column("number1", NativeTypes.numberOf(19), false),
+                new Column("number2", NativeTypes.numberOf(10), false)
+            }
+        );
+
+        final TupleBuilderImpl tup = new TupleBuilderImpl(schema);
+
+        tup.set("key", rnd.nextLong());
+        tup.set("number1", pair.getFirst());
+        tup.set("number2", pair.getSecond());
+
+        Tuple keyTuple = new TupleBuilderImpl(schema).set("key", rnd.nextLong()).build();
+
+        final Row row = marshaller.marshal(keyTuple, tup.build());
+
+        assertEquals(row.numberValue(1), row.numberValue(2));
+    }
+
+    /**
+     *
+     */
+    @Test
+    public void testPrecisionRestrictionsForNumbers() {
+        schema = new SchemaDescriptor(
+            UUID.randomUUID(),
+            42,
+            new Column[] {new Column("key", NativeTypes.INT64, false)},
+            new Column[] {
+                new Column("number1", NativeTypes.numberOf(5), false),
+            }
+        );
+
+        final TupleBuilderImpl badTup = new TupleBuilderImpl(schema);
+
+        badTup.set("key", rnd.nextLong());
+        assertThrows(InvalidTypeException.class, () -> badTup.set("number1", BigInteger.valueOf(999991L)), "Column's type mismatch");
+        assertThrows(InvalidTypeException.class, () -> badTup.set("number1", new BigInteger("111111")), "Column's type mismatch");
+        assertThrows(InvalidTypeException.class, () -> badTup.set("number1", BigInteger.valueOf(-999991L)), "Column's type mismatch");
+        assertThrows(InvalidTypeException.class, () -> badTup.set("number1", new BigInteger("-111111")), "Column's type mismatch");
+    }
+
+    /**
+     *
+     */
+    @Test
+    public void testPrecisionRestrictionsForDecimal() {
+        schema = new SchemaDescriptor(
+            UUID.randomUUID(),
+            42,
+            new Column[] {new Column("key", NativeTypes.INT64, false)},
+            new Column[] {
+                new Column("decimalCol", NativeTypes.decimalOf(12, 3), false),
+            }
+        );
+
+        final TupleBuilderImpl badTup = new TupleBuilderImpl(schema);
+
+        badTup.set("key", rnd.nextLong());
+        assertThrows(InvalidTypeException.class, () -> badTup.set("decimalCol", new BigDecimal("123456789.0123")), "Column's type mismatch");
+        assertThrows(InvalidTypeException.class, () -> badTup.set("decimalCol", new BigDecimal("-1234567890123")), "Column's type mismatch");
+        assertThrows(InvalidTypeException.class, () -> badTup.set("decimalCol", new BigDecimal(123456789.0123d)), "Column's type mismatch");
```

Comment:
Let's add a test for 123.123456789 and 123456789.3210

Reply:
I added lot`s of cases with different bigDecimal representation, as well as yours.